main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright 2017 Jason Mittertreiner
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.Symbol.EnumSym
+import ca.uwaterloo.flix.language.ast.{Type, TypedAst}
+import ca.uwaterloo.flix.language.ast.TypedAst.Pattern
+import ca.uwaterloo.flix.language.ast.TypedAst.Pattern.Tag
+import ca.uwaterloo.flix.language.errors.ExhaustiveMatchError
+import ca.uwaterloo.flix.language.{CompilationError, GenSym}
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation._
+
+/**
+  * The Pattern Exhaustiveness phase checks pattern matches for exhaustiveness
+  * as well as for useless patterns
+  *
+  * A pattern is useless if:
+  * A pattern match is exhaustive if:
+  */
```

Comment:
We probably a general description of the overall algorithm and a link to the paper.